tensorflow/lite/experimental/shlo/ops/convolution.cc

```diff
@@ -431,6 +433,143 @@ absl::Status PrepareImpl(ConvolutionOp& op, const Tensor& lhs,
+    lhs_padded_spatials[i - 2] = lhs_transposed.shape().Dim(i) +
+                                 (op.attributes.lhs_dilation[i - 2] - 1) *
+                                     (lhs_transposed.shape().Dim(i) - 1) +
+                                 padding_buffer[2 * (i - 2)] +
+                                 padding_buffer[(2 * (i - 2)) + 1];
+    lhs_padded_tensor_size *= lhs_padded_spatials[i - 2];
+  }
+
+  lhs_padded_tensor_size *=
+      lhs_transposed.shape().Dim(0) * lhs_transposed.shape().Dim(1);
+  op.lhs_padded_data =
+      std::vector<std::byte>(lhs_padded_tensor_size * sizeof(StorageT));
+  absl::InlinedVector<DimensionSize, kMaxNumDimensions> lhs_padding_shape_dims(
+      lhs_transposed.Rank(), 0);
+  lhs_padding_shape_dims[0] = lhs_transposed.shape().Dim(0);
+  lhs_padding_shape_dims[1] = lhs_transposed.shape().Dim(1);
+  for (size_t i = 0; i < lhs_transposed.Rank() - 2; ++i) {
+    lhs_padding_shape_dims[i + 2] =
+        static_cast<int64_t>(lhs_padded_spatials[i]);
+  }
+  const Shape lhs_padding_shape(lhs_padding_shape_dims);
+  Tensor lhs_padded{.type = TensorType{.shape = lhs_padding_shape,
+                                       .element_type = storage_type},
+                    .data = op.lhs_padded_data.data()};
+  int64_t pad_output_shape[kMaxNumDimensions];
+  std::copy(lhs_padding_shape_dims.data(),
+            lhs_padding_shape_dims.data() + lhs_transposed.Rank(),
+            pad_output_shape);
+  int64_t edge_pad_high[kMaxNumDimensions];
+  int64_t edge_pad_low[kMaxNumDimensions];
+  int64_t interior_pad[kMaxNumDimensions];
+  edge_pad_high[0] = edge_pad_low[0] = interior_pad[0] = 0;
+  edge_pad_high[1] = edge_pad_low[1] = interior_pad[1] = 0;
+  for (int64_t i = 2; i < lhs_transposed.Rank(); ++i) {
+    edge_pad_low[i] = padding_buffer[2 * (i - 2)];
+    edge_pad_high[i] = padding_buffer[2 * (i - 2) + 1];
+    interior_pad[i] = op.attributes.lhs_dilation[i - 2] - 1;
+  }
+  int64_t pad_rank = lhs_transposed.Rank();
+  int64_t pad_output_dimension_sizes[kMaxNumDimensions];
+  pad_output_dimension_sizes[pad_rank - 1] = 1;
+  op.pad_output_strides.resize(pad_rank, 0);
+  op.pad_output_strides[pad_rank - 1] = interior_pad[pad_rank - 1] + 1;
+  for (int64_t i = pad_rank - 2; i >= 0; --i) {
+    pad_output_dimension_sizes[i] =
+        pad_output_shape[i + 1] * pad_output_dimension_sizes[i + 1];
+    op.pad_output_strides[i] =
+        pad_output_dimension_sizes[i] * (interior_pad[i] + 1);
+  }
+  for (int64_t i = 0; i < pad_rank; ++i) {
+    op.pad_output_offset +=
+        std::max<int64_t>(edge_pad_low[i], 0) * pad_output_dimension_sizes[i];
+  }
+  op.pad_input_strides.resize(pad_rank, 0);
+  op.pad_input_strides[pad_rank - 1] = 1;
+  for (int64_t i = pad_rank - 1; i >= 1; --i) {
+    op.pad_input_strides[i - 1] =
+        lhs_transposed.shape().Dim(i) * op.pad_input_strides[i];
+  }
+  auto DivNegRoundAwayOrZero = [](int64_t num, int64_t denum) -> int64_t {
```

Comment:
we are not using lambda functions in this library, so make it as a standalone function in helper_functions.h file